lib/base/tlsstream.cpp

```diff
@@ -103,3 +104,45 @@ void UnbufferedAsioTlsStream::BeforeHandshake(handshake_type type)
 	}
 #endif /* SSL_CTRL_SET_TLSEXT_HOSTNAME */
 }
+
+/**
+ * Forcefully close the connection, typically (details are up to the operating system) using a TCP RST.
+ */
+void AsioTlsStream::ForceDisconnect()
+{
+	boost::system::error_code ec;
+
+	// Close the socket. In case the connection wasn't shut down cleanly by GracefulDisconnect(), the operating system
+	// will typically terminate the connection with a TCP RST. Otherwise, this just releases the file descriptor.
+	lowest_layer().close(ec);
+}
+
+/**
+ * Try to cleanly shut down the connection. This involves sending a TLS close_notify shutdown alert and terminating the
+ * underlying TCP connection. Sending these additional messages can block, hence the method takes a yield context and
+ * internally implements a timeout of 10 seconds for the operation after which the connection is forcefully terminated
+ * using ForceDisconnect().
+ *
+ * @param strand Asio strand used for other operations on this connection.
+ * @param yc Yield context for Asio coroutines
+ */
+void AsioTlsStream::GracefulDisconnect(boost::asio::io_context::strand strand, boost::asio::yield_context yc)
+{
+	boost::system::error_code ec;
+
+	Timeout::Ptr shutdownTimeout(new Timeout(strand.context(), strand, boost::posix_time::seconds(10),
+		[this, keepAlive = AsioTlsStream::Ptr(this)](boost::asio::yield_context yc) {
+			// Forcefully terminate the connection if async_shutdown() blocked more than 10 seconds.
+			ForceDisconnect();
```

Comment:
Why not use `lowest_layer().cancel()` here as before? Shouldn't this call force `async_shutdown` to return? Right now, if you force disconnect here, which basically means you literally destroy the socket file descriptor, the call to `shutdown()` on the underlying TCP layer becomes pointless since the socket no longer exists at that point. I know we suppress the errors, but `shutdown()` would return a `bad file descriptor` error in this case, which I don't think is right.

Reply:
> And second, if the socket is closed, it prevents accidentally starting new operations.

Even then, shouldn't the last `ForceDisconnect()` call be enough?

> Adding a few calls to `is_open()` should not hurt though, that would hopefully make this a bit more obvious.

Instead of adding such conditions, I would simply use `cancel()` here as before, as you are going to destroy the socket at the end of the coroutine anyway.

Reply:
Ah, the discussion you had in https://github.com/Icinga/icinga2/pull/10005#discussion_r1844132097 is also relevant here. The check in the code I linked works because `close()` sets the internal file descriptor to -1 (which is the value of `invalid_socket`).